File: src/execution/runtimeFederationEngine.ts
/**
 * Converted from Python: core/execution/runtime_federation_engine.py
 * @generated — WebWeaveX python→javascript library port
 */

import * as py from "../runtime/pyCompat.js";
import { buildRuntimeWorkers } from "./runtimeWorkerEngine.js";

export function federateRuntimeExecution(workers: any, actions: any = null): any {
  var built: any = buildRuntimeWorkers(workers);
  var routes: any[] = [];
  var index: any;
  var worker: any;
  for ([index, worker] of py.enumerate(built)) {
    var action: any = (py.truthy(actions) ? py.at(py.or2(actions, () => ([{}])), py.mod(index, py.max([py.len(py.or2(actions, () => ([{}]))), 1]))) : {});
    py.listAppend(routes, {"worker_id": py.at(worker, "worker_id"), "runtime": py.at(worker, "runtime"), "action_id": py.toStr(py.get(action, "id", `route:${py.toStr(index)}`)), "route_order": index});
  }
  return {"workers": built, "execution_routes": py.sorted(routes, {key: ((item: any) => py.at(item, "route_order")) as (item: any) => any}), "federated": true, "bounded": true};
}
export { buildRuntimeWorkers };

File: lib/src/execution/runtime_federation_engine.dart
import 'runtime_worker_engine.dart';

/// Port of core/execution/runtime_federation_engine.federate_runtime_execution.
Map<String, dynamic> federateRuntimeExecution(
  List<Map<String, dynamic>> workers, [
  List<Map<String, dynamic>>? actions,
]) {
  final List<Map<String, dynamic>> built = buildRuntimeWorkers(workers);
  final List<Map<String, dynamic>> routes = <Map<String, dynamic>>[];

  // Mirrors Python: actions or [{}] for the modulo index; if actions falsy
  // (None or empty), the per-route action is {} entirely.
  final bool hasActions = actions != null && actions.isNotEmpty;
  final List<Map<String, dynamic>> actionPool =
      hasActions ? actions : <Map<String, dynamic>>[<String, dynamic>{}];

  for (int index = 0; index < built.length; index++) {
    final Map<String, dynamic> worker = built[index];
    final Map<String, dynamic> action = hasActions
        ? actionPool[index % (actionPool.isEmpty ? 1 : actionPool.length)]
        : <String, dynamic>{};
    routes.add(<String, dynamic>{
      'worker_id': worker['worker_id'],
      'runtime': worker['runtime'],
      'action_id': '${action['id'] ?? 'route:$index'}',
      'route_order': index,
    });
  }

  routes.sort((Map<String, dynamic> a, Map<String, dynamic> b) =>
      (a['route_order'] as int).compareTo(b['route_order'] as int));

  return <String, dynamic>{
    'workers': built,
    'execution_routes': routes,
    'federated': true,
    'bounded': true,
  };
}
